src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -1039,50 +1039,82 @@ private void RemoveWorseMembers<TMember>(ArrayBuilder<MemberResolutionResult<TMe
+                    worse[index] = betterThanEverything;
+                    currentBestIndex = index;
+                }
+                else
                 {
-                    var c2result = results[c2Idx];
+                    worse[currentBestIndex] = notBetterThanEverything;
+                    worse[index] = notBetterThanEverything;
 
-                    if (!c2result.Result.IsValid)
+                    while (++currentBestIndex < results.Count)
                     {
-                        continue;
+                        if (results[currentBestIndex].Result.IsValid)
+                        {
+                            worse[currentBestIndex] = betterThanEverything;
+                            break;
+                        }
                     }
 
-                    if (c1result.Member == c2result.Member)
-                    {
-                        continue;
-                    }
+                    index = currentBestIndex;
+                }
+            }
+
+            for (int index = 0; index < results.Count; index++)
+            {
+                if (index == currentBestIndex || !results[index].Result.IsValid)
+                {
+                    continue;
+                }
 
-                    var better = BetterFunctionMember(c1result, c2result, arguments.Arguments, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
+                var better = BetterFunctionMember(results[currentBestIndex], results[index], arguments.Arguments, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    worse[index] = worseThanSomething;
+                }
+                else if (better == BetterResult.Right)
+                {
+                    worse[currentBestIndex] = worseThanSomething;
+                    break;
+                }
+                else
+                {
+                    if (worse[currentBestIndex] != worseThanSomething)
                     {
-                        worse[c2Idx] = worseThanSomething;
+                        worse[currentBestIndex] = notBetterThanEverything;
                     }
-                    else
+
+                    if (worse[index] != worseThanSomething)
                     {
-                        c1IsBetterThanEverything = false;
-                        if (better == BetterResult.Right)
-                        {
-                            worse[c1Idx] = worseThanSomething;
-                            break;
-                        }
+                        worse[index] = notBetterThanEverything;
                     }
-                }
-                if (worse[c1Idx] == unknown)
-                {
-                    // c1 was not worse than anything. Was it better than everything?
-                    worse[c1Idx] = c1IsBetterThanEverything ? betterThanEverything : notBetterThanEverything;
+
+                    break;
```

Comment:
The risk of this `break` is that we may have some candidate marked `worse[i] == notBetterThanEverything` by the old code and `worseThanSomething` by the new code. That would cause it to be handled incorrectly later. I think.